src/gsi/core_disp_func.F90

```diff
@@ -0,0 +1,1055 @@
+!!$         x*(3.0*(f(0) - f(1)) + f(2) - f(-1))))
+
+  cubicInterpolate = f(0) + &
+       x     * 0.5 * (f(1) - f(-1)) + &
+       x*x   * 0.5 * (2.0*f(-1) - 5.0*f(0) + 4.0*f(1) - f(2)) + &
+       x*x*x * 0.5 * (3.0*(f(0) - f(1)) + f(2) - f(-1))
+
+end function cubicInterpolate
+
+real(fp) function bilin_interp(f00, f10, f01, f11, dx, dy)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram:    bilin_interp
+!   prgmmr: Nehrkorn
+!
+! abstract: bilinear interpolation of 4 nearest neighbors to a point
+!
+! program history log:
+!
+!   input argument list:
+!     f00, f10, f01, f11, dx, dy
+!
+!   output argument list:
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+    ! bilinear interpolation of 4 nearest neighbors to a point
+    implicit none
+    ! bilinear interpolate in the unit box.
+    real(fp), intent(in) :: f00, f10, f01, f11, dx, dy
+    bilin_interp = f00*((1.-dx)*(1.-dy)) + f10*(dx*(1.-dy)) + f01*((1.-dx)*dy) + f11*(dx*dy)
+    return
+end function bilin_interp
+
+subroutine compute_xy_orig(x_disp, y_disp, ix_orig, iy_orig, dx_n, dy_n, meta, order)
+!$$$  subprogram documentation block
+!                .      .    .                                       .
+! subprogram: compute_xy_orig
+!   prgmmr: Nehrkorn
+!
+! abstract: apply displacements to a gridded 2d field
+!  NOTE: gridded fields assumed to be on the same grid as the displacements
+!
+! program history log:
+!
+!   input argument list: original field in bg, displacements in disp
+!     x_disp, y_disp,  dx_n, dy_n
+!     ix_orig, iy_orig
+!     meta
+!     order
+!
+!   output argument list: displaced field in adj
+!     x_disp, y_disp,  dx_n, dy_n, dx_n, dy_n
+!     ix_orig, iy_orig
+!
+! attributes:
+!   language: f90
+!   machine:
+!
+!$$$ end documentation block
+  implicit none
+  real(fp), dimension(:,:), intent(inout) :: x_disp, y_disp, dx_n, dy_n
+  integer, dimension(:,:), intent(inout) :: ix_orig, iy_orig
+  type (fca_gridded_meta), intent(in) :: meta		! meta data
+  integer, intent(in) :: order
+
+  real(fp), dimension(size(x_disp,1),size(x_disp,2)) :: x, y
+  integer :: i, j, status, buf, Nx, Ny
+
+#ifdef TRACE_USE
+  if (trace_use) call da_trace_entry("compute_xy_orig")
+#endif
+
+  Nx = size(x_disp,1)
+  Ny = size(x_disp,2)
+  forall(i=1:Nx,j=1:Ny) 
+     x(i,j) = its-1+(i-1)*meta%dx + x_disp(i,j) + meta%x0
+     y(i,j) = jts-1+(j-1)*meta%dy + y_disp(i,j) + meta%y0
+  end forall
+
+  ! new array indices.
+
+  ix_orig(1:Nx,1:Ny) = int(floor(x(1:Nx,1:Ny)/meta%dx)+1) ! x index of lower-left corner of containing value to bring back here
+  iy_orig(1:Nx,1:Ny) = int(floor(y(1:Nx,1:Ny)/meta%dy)+1) ! y ""
+
+  ! use constant extrapolation for points outside the grid: 
+  ! (Note that since ide/jde is padded by one for staggered grids, we need to use ide-1 in the comparison below)
+
+  if (order .eq. bilinear) then
```

Comment:
Please replace .eq. with ==

Reply:
The logic operators in the AER code have been updated as requested.